community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/storeview/LabelScanViewNodeStoreScan.java

```diff
@@ -0,0 +1,211 @@
+            PrimitiveLongSet nodes = labelNodes.get( labelId );
+            if ( nodes != null )
+            {
+                long[] labels = new long[]{labelId};
+                PrimitiveLongIterator nodeIterator = nodes.iterator();
+                NodeRecord nodeRecord = new NodeRecord( -1 );
+                try ( IndexUpdater updater = indexPopulator.newPopulatingUpdater( storeView ) )
+                {
+                    while ( nodeIterator.hasNext() )
+                    {
+                        long nodeId = nodeIterator.next();
+                        NodeRecord record = nodeStore.getRecord( nodeId, nodeRecord, RecordLoad.FORCE );
+
+                        int propertyKeyId = descriptor.getPropertyKeyId();
+
+                        updater.process(
+                                NodePropertyUpdate.remove( nodeId, propertyKeyId, StringUtils.EMPTY, labels ) );
+                        if ( record.inUse() )
+                        {
+                            Property property = storeView.getProperty( nodeId, propertyKeyId );
+                            if ( property.isDefined() )
+                            {
+                                Object propertyValue = property.value();
+                                updater.process(
+                                        NodePropertyUpdate.add( nodeId, propertyKeyId, propertyValue, labels ) );
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Checks if provided node property update is applicable for current store view.
+     * In case if we use label store view - updates are always applicable, otherwise fallback to default behaviour of
+     * neo store view.
+     *
+     * @param updater updater that should process updates in required
+     * @param update node property update
+     * @param currentlyIndexedNodeId currently indexed node id
+     */
+    @Override
+    public void acceptUpdate( MultipleIndexPopulator.MultipleIndexUpdater updater, NodePropertyUpdate update,
+            long currentlyIndexedNodeId )
+    {
+        long[] labels;
+        switch ( update.getUpdateMode() )
+        {
+        case ADDED:
+            labels = update.getLabelsAfter();
+            break;
+        case CHANGED:
+            labels = Arrays.stream( update.getLabelsBefore() )
+                    .filter( item -> Arrays.binarySearch( update.getLabelsAfter(), item ) >= 0 )
+                    .toArray();
+            break;
+        case REMOVED:
+            labels = update.getLabelsBefore();
+            break;
+        default:
+            throw new UnsupportedOperationException( "unsupported update mode" );
+        }
+        int property = update.getPropertyKeyId();
+        long nodeId = update.getNodeId();
+        for ( long label : labels )
+        {
+            PrimitiveLongObjectMap<PrimitiveLongSet> labelsNodes = propertyLabelNodes.get( property );
```

Comment:
Another thing which may be good would be to, instead of queueing up _all_ updates during the scan, to look at number of pending updates and if they are rising then the label scan could do this incrementally: i.e. at some point during the scan stop, apply pending updates and create a new reader, resetting the update queue.